transform_fix.ml

```diff
@@ -15,44 +15,54 @@ let remove_falltrough ({instrs} as inp : analysis_input) =
     in
     match[@warning "-4"] instrs.(pc) with
     | Label (MergeLabel l) ->
-        if has_fallthrough l pc
+        if has_fallthrough pc
         then InsertBeforeLabel [ Goto l ]
         else Unchanged
     | Label (BranchLabel l) ->
         let l' = fresh_label l (string_of_int pc) in
-        if has_fallthrough l pc
-        then ReplaceLabel [ Goto l'; Label (BranchLabel l); Goto l'; Label (MergeLabel l') ]
+        if has_fallthrough pc
+        then ReplaceLabel [ Goto l';
+                            Label (BranchLabel l); Goto l';
+                            Label (MergeLabel l') ]
         else Unchanged
     | _ -> Unchanged
   in
   change_instrs normalize inp
 
 let make_branch_targets_unique ({instrs} as inp : analysis_input) =
   let normalize pc =
-    let incomming_branches label =
+    let incoming_branches label =
       (* should not rely on preds on a broken graph *)
-      let rec incomming pc =
+      let rec incoming pc =
         if pc = Array.length instrs then [] else
         match[@warning "-4"] instrs.(pc) with
-        | Branch (_, l1, l2) when l1 = label || l2 = label -> pc :: incomming (pc+1)
-        | _-> incomming (pc+1)
+        | Branch (_, l1, l2) when l1 = label && l2 = label ->
+            pc :: pc :: incoming (pc+1)
+        | Branch (_, l1, l2) when l1 = label || l2 = label ->
+            pc :: incoming (pc+1)
+        | _-> incoming (pc+1)
       in
-      incomming 0
+      incoming 0
     in
-    let fresh_label label tag = (Edit.fresh_label instrs label) ^ "_" ^ tag in
+    let fresh_label label tag = (Edit.fresh_label instrs (label ^ "_" ^ tag)) in
     match[@warning "-4"] instrs.(pc) with
     | Label (BranchLabel l) ->
-        if List.length (incomming_branches l) > 1
+        if List.length (incoming_branches l) > 1
         then ReplaceLabel [ Label (MergeLabel l) ]
         else Unchanged
     | Branch (e, l1, l2) ->
-        let i1, i2 = incomming_branches l1, incomming_branches l2 in
+        let i1, i2 = incoming_branches l1, incoming_branches l2 in
         begin match i1, i2 with
         | _, [] -> assert(false)
         | [], _ -> assert(false)
         | [pc1], [pc2] ->
             assert(pc1 = pc && pc2 = pc);
-            Unchanged
+            if l1 <> l2 then Unchanged else
```

Comment:
Still a bit confused about this, sorry.

If `incoming_branches` returned singleton lists, then `l1` and `l2` have to be different. If `l1` and `l2` are the same, then because of lines 39-40 above, `incoming_branches` would return a list with at least two `pc`s.